### PR TITLE
Fixes a bug in dry size calculations and these calculations now use current state

### PR DESCRIPTION
**Pull request by singhbalwinder**
Dry aerosol size calculations were using an incorrect log of 
sigma g (geometric std. dev. of aerosol mode) during
accumulation<->aitken transfer. This Pr fixes that bug.

This PR is NBFB as dry aerosol sizes and water uptake are 
now using the current "state" (mass mixing ratios of the 
species). Earlier, an older "state" was used to compute
these quantities. I have modified RRTMGP and CRM codes
to accommodate this change.

An extensive refactor of modal_aero_calcsize is not included
in this PR so that it is easier to understand where the NBFB
differences are coming from. Code in this branch is now
BFB with the code where modal_aero_calcsize is refactored
extensively. I will port those changes as a separate PR which
will be BFB.

[NBFB]

### CODE REVIEW

**Comment by singhbalwinder** [comment]
I have run RRTMGP and CRM tests from the integration test suite and they work fine with these changes.

I will revisit these codes in a separate PR to clean up some of the logic. I hope to remove calsize_diag and have just calcsize_sub routine to work as calcsize_diag for prognostic aerosol and diagnostic calculations.

**Comment by singhbalwinder** [comment]
Here are the diagnostics showing impact of this bug fix and having the calcsize use the current state:
https://compy-dtn.pnl.gov/sing201/diags/BugfixedCalcsize_vs_Master/viewer/

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (whannah1) [inline components/eam/src/physics/cam/modal_aer_opt.F90:1320]: Don't we need logic for when clear_rh is present here?

**Review by whannah1 — APPROVED** [review]
This looks good to me as it is, but It's a shame that we have to duplicate so many lines for the clear_rh to be passed into the radiation. I vaguely recall being able to just pass an optional argument down and the state of being "present" or not travels with that argument, so we might not need all these new "if present" blocks.

File: components/eam/src/chemistry/utils/modal_aero_calcsize.F90
Comment (singhbalwinder) [reply components/eam/src/chemistry/utils/modal_aero_calcsize.F90:646]: It probably doesn't matter here but sometime, it can make a difference. It also makes it more explicit that `tadj` is a double precision real number.

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (singhbalwinder) [reply components/eam/src/physics/cam/modal_aer_opt.F90:1320]: I will clean this up in my next BFB PR. Since MMF was not using prognostic aerosols, I didn't include this logic for `clear_rh`

**Comment by kaizhangpnl** [comment]
@singhbalwinder Thanks again for working on this. Does this PR also resolve the memory problem we had before?

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (kaizhangpnl) [inline components/eam/src/physics/cam/modal_aer_opt.F90:405]: Would it be better to use "clear_rh_mmf" here, so that we know this is for MMF?

File: components/eam/src/chemistry/utils/modal_aero_calcsize.F90
Comment (kaizhangpnl) [reply components/eam/src/chemistry/utils/modal_aero_calcsize.F90:1173]: Will your comment here be permanent? If not, it would be better to have the bugfix information (date and reason) in the code.

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (singhbalwinder) [reply components/eam/src/physics/cam/modal_aer_opt.F90:405]: I am revisiting this logic in a follow-up PR. I will change the name of this variable and add a comment about it in that PR.

**Comment by singhbalwinder** [comment]
@kaizhangpnl : My comment about bug fix is permanent but only on GitHub. I will add a comment about this in the code in a follow up PR. Thanks for your review.

**Review by kaizhangpnl — APPROVED** [review]

**Comment by singhbalwinder** [comment]
> @singhbalwinder Thanks again for working on this. Does this PR also resolve the memory problem we had before?

Sorry, I missed answering this question. The memory issue is more apparent when we invoke diagnostics (which are disabled here). This PR partly addresses that issue by allocating arrays in the init routine rather than at each time step.

**Comment by wlin7** [comment]
Hi @singhbalwinder , should we wait for approval from all the suggested reviewers for this PR?

**Review by hwangacme — APPROVED** [review]

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (brhillman) [reply components/eam/src/physics/cam/modal_aer_opt.F90:1320]: Regardless the answer is "no". The "presence" of an optional parameter like `clear_rh` would be passed to a subroutine. So actually I think this whole block can be cleaned up by just passing `clear_rh_in=clear_h` in both cases. Nothing wrong with being explicit though probably, so long as the clear_rh present case gets fixed.

File: components/eam/src/physics/rrtmgp/cam_optics.F90
Comment (brhillman) [inline components/eam/src/physics/rrtmgp/cam_optics.F90:510]: I believe these logical blocks are unnecessary, you should just be able to pass clear_rh and avoid the duplicated code.

File: components/eam/src/physics/cam/aer_rad_props.F90
Comment (brhillman) [inline components/eam/src/physics/cam/aer_rad_props.F90:268]: I think this extra logic is unnecessary; you can just pass clear_rh as an optional and `model_aero_sw` will determine if it is a valid argument.

File: components/eam/src/physics/cam/aer_rad_props.F90
Comment (brhillman) [inline components/eam/src/physics/cam/aer_rad_props.F90:432]: Again should just be able to pass clear_rh.

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (brhillman) [inline components/eam/src/physics/cam/modal_aer_opt.F90:622]: Can just pass clear_rh to both.

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (brhillman) [inline components/eam/src/physics/cam/modal_aer_opt.F90:1309]: Again could just pass clear_rh in both calls.

**Review by brhillman — APPROVED** [review]
This looks fine to me. As @whannah1 said, the additional logic around calls that pass `clear_rh` are unnecessary; you should be able to pass that optional argument in all cases, and the called routine will check whether the optional argument is a valid parameter. That is, the behavior of "present" should follow from the top down. The advantage of the extra logic you've added is that it's explicit, but the disadvantage is that I think it hurts the readability and maintainability a little, because of the duplicated code blocks. That said, I'm fine if you want to leave it as is in this PR.

**Comment by singhbalwinder** [comment]
Thanks @hwangacme and @brhillman for your reviews.

I agree that the "present" logic is a little messy. This part of the code is very tricky as there are multiple versions of radiation.F90 in the code (for RRTMG, RRTMGP, CAMRT, CAM etc.). This makes it really hard to predict which version will be called in a particular configuration (normal, RRTMGP, MMF etc.). I placed these "present" logic so that the code works in all scenarios. I will see if I can remove some of these in my followup PR. If the "present" logic is missing and the optional variable is accessed, the code generally crashes with a segmentation fault (without much useful info even in debug mode). Since `clear_rh` is only provided by MMF codes, I will see if I can remove the "present" logic in routines never called by MMF codes.

@wlin7 : Please go ahead and start merging this PR if repo is open. I will rebase my followup PR once this is merged and address all these issues in that PR.

**Comment by brhillman** [comment]
@singhbalwinder I didn't mean to remove the present logic in the low level routines, I mean that you can remove the present logic around the subroutine calls, because you can pass an optional parameter clear_rh whether it evaluates present or not. Passing an optional that may or may not be there is different than accessing the variable directly, which I agree would create a segfault.

BTW, we have worked pretty hard to keep the call signatures as consistent as possible between RRTMGP, RRTMG, and the MMF version of RRTMGP. I'm not sure how I feel about them diverging. Walter's clear_rh addition may even be useful outside of the MMF context, I don't know. I haven't looked closely enough at it.

**Comment by wlin7** [comment]
Merged to next.

**Comment by wlin7** [comment]
Merged to master.

Request to bless 44 DIFF sandiatoss3 next tests, leave out pre-existing 2 DIFFs
Request to bless 40 DIFF compy integration_next tests. leave out 1 pre-existing DIFF; compy has 3 new FAIL, one of which was DIFF in the previous return.

 SMS_D_Ld1.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.compy_pgi.allactive-v1cmip6
 SMS_Ld9.ne4pg2_ne4pg2.FC5AV1C-04P2.compy_pgi.eam-thetahy_sl_pg2_mass
 SMS_D_Ld1.ne30_r05_oECv3.A_WCYCL1850.compy_pgi  (DIFF to FAIL)

To watch for test returns for other machines. behop all tests failed, there must be other systematic issue.

**Comment by wlin7** [comment]
Bless 11 tests for mappy. Note that while tested using the same commit, 6 DIFFs reported on 10/15/2020, but 11 on 10/16/2020. DIFFs on 10/15/2020 were not blessed.

**Comment by wlin7** [comment]
Request to bless 3 tests each for anvil and compy e3sm_prod_next_intel.

Also bless compy e3sm_integration_next_pgi on 2 tests that were  FAILs since the merge of this PR until 10/21/20. Both passed before this PR. Presumably the current DIFFs (as of 10/22/20) are due to this PR.

```
SMS_D_Ld1.ne30_oECv3_ICG.A_WCYCL1850S_CMIP6.compy_pgi.allactive-v1cmip6
SMS_Ld9.ne4pg2_ne4pg2.FC5AV1C-04P2.compy_pgi.eam-thetahy_sl_pg2_mass
```

Two other DIFFs in compy e3sm_integration_next_pgi existed well before this PR and remain so. Although the current PR contributes to DIFF, they are yet to be blessed. Listed below for reference

```
MS_D_Ld1.ne30_r05_oECv3.A_WCYCL1850.compy_pgi 
SMS_Ld1.ne4_ne4.F-EAMv1-AQP1.compy_pgi.eam-clubb_only


```

**Comment by wlin7** [comment]
@rljacob , what do we need to do about the remaining two DIFFs (see comment immediately above) in compy e3sm_integration_next_pgi?

**Comment by rljacob** [comment]
Go ahead and request a bless for them.  They were probably missed earlier.

**Comment by wlin7** [comment]
Request submitted to bless the two remaining compy DIFFs as noted above.

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (ambrad) [inline components/eam/src/physics/cam/modal_aer_opt.F90:75]: @singhbalwinder should wetdens_m appear in the threadprivate list?

Comment (singhbalwinder) [reply components/eam/src/physics/cam/modal_aer_opt.F90:75]: `wetdens` is not being used in the code (or impact any calculations) and I have removed it entirely in my revised code (PR pending). 

I created this PR (with minimal changes)so that we can get all the NBFB changes in the master. Sorry about the confusion.

File: components/eam/src/physics/cam/modal_aer_opt.F90
Comment (ambrad) [inline components/eam/src/physics/cam/modal_aer_opt.F90:621]: @singhbalwinder there might be another potential issue, if I've traced the code correctly.

1. Here, dgnumdry_m is passed to modal_aero_calcsize_diag.
2. dgnumdry_m was allocated during initialization as a module-level array that persists throughout the run.
3. But modal_aero_calcsize_diag calls `allocate(dgnum_m(pcols,pver,nmodes), stat=stat)` for the given set of arguments to modal_aero_calcsize_diag that we see at this line.
4. Thus, it appears that dgnumdry_m is allocated at each call to modal_aero_calcsize_diag. But because this module assumes dgnumdry_m is allocated just once at initialization, these subsequent allocations are not deallocated. And since dgnum_m has type `pointer` rather than `allocatable`, it won't be auto-deallocated.

Sorry if I got that wrong.

Comment (singhbalwinder) [reply components/eam/src/physics/cam/modal_aer_opt.F90:621]: Thanks Andrew. This indeed seems like a bug. My goal was to remove “calcsize_diag” entirely and replace it with “calcsize_sub” but there isn’t enough time now to remove it and test it throughly. I was testing “calcsize_sub” routine all the time because it is one relevant for prognostic aerosols. “calcsize_diag” is never called for prognostic aerosols now.

This routine is called from different places in the code with different configurations. My intention here was to get dry size (dgnum) using the "fresh state” even with list_dix = 0. The extra allocate should be removed and we should expect dgnum to be already allocated in this routine. Thanks again for catching this!

**Pull request merged**
